gigaword-indexer/src/main/java/edu/isi/vista/gigawordIndexer/ConcatenatedGigawordDocuments.java

```diff
@@ -18,18 +26,112 @@
+    this.concatenatedFileText = Objects.requireNonNull(concatenatedFileText);
+    this.articleList = aArticleList;
+    this.annotated = true;
+  }
+  
   public static ConcatenatedGigawordDocuments fromGigwordGZippedFile(Path p) throws IOException {
     return new ConcatenatedGigawordDocuments(GZIPByteSource.fromCompressed(p.toFile())
             .asCharSource(StandardCharsets.UTF_8).read());
   }
 
+  public static ConcatenatedGigawordDocuments fromAnnotatedGigwordGZippedFile(Path p) throws IOException {
+    // Initialize Streams
+    try (InputStream fileInputStream = new FileInputStream(p.toString());
+         GZIPInputStream gzipInputStream = new GZIPInputStream(fileInputStream);
+         InputStreamReader inputStreamReader = new InputStreamReader(gzipInputStream);
+         BufferedReader br = new BufferedReader(inputStreamReader)) {
+
+        String sCurrentLine;
+    
+        // variables for start of a new document
+        Pattern GIGAWORD_DOC_ELEMENT_PATTERN = Pattern.compile("<DOC id=\"(.*?)\".*>");
+        String currentDocId = "";
+    
+        // variables for saving articles
+        boolean headlineStarted = false;
+        boolean textStarted = false;
+        String docText = "";
+        List<Article> articleList = new ArrayList<>();
+    
+    
+        // read file
+        while ((sCurrentLine = br.readLine()) != null) {
+    
+            sCurrentLine = sCurrentLine.trim();
+    
+            // extract document ID
+            if (sCurrentLine.startsWith("<DOC id=")) {
+                Matcher m = GIGAWORD_DOC_ELEMENT_PATTERN.matcher(sCurrentLine);
+                if (m.find()) {
+                    currentDocId = m.group(1);
+                }
+                else {
+                    throw new RuntimeException("Missing document ID on article");
+                }
+            }
+    
+            // create article when end of document (</TEXT>) is reached
+            if (sCurrentLine.equals("</TEXT>")) {
+                articleList.add(new Article(currentDocId, docText));
+                docText = "";
+                textStarted = false;
+            }
+    
+            if (sCurrentLine.equals("</HEADLINE>")) {
+                docText += "\n\n";
+                headlineStarted = false;
+            }
+    
+            // read lines from document text
+            if ((headlineStarted || textStarted) && !sCurrentLine.startsWith("<")) {
+                if (sCurrentLine.equals("")) {
+                    docText += "\n";
+                }
+                else if (sCurrentLine.startsWith("(")) {
+                    String parsedText = PennTreeUtils.toText(PennTreeUtils.parsePennTree(sCurrentLine));
+                    if (!parsedText.equals("null")) {
+                        docText += parsedText + " ";
+                    }
+                }
+                // sometimes there is one unannotated line in the document which can be used as plain text
```

Comment:
When does this happen?

Reply:
I really don't see a pattern. It almost seems as if some sentences just were forgotten.
Annotated sentences are usually between <P> and </P> tags, and these unannotated sentences are as well.

Reply:
Here's an example:

```
<P>
( (S (NP (NNP Zito)) (ADVP (RB also)) (VP (VP (MD will) (VP (VB send) (PRT (RP out)) (NP (PRP It) (VBZ 's)) (PP (IN So) (NP (NP (NN Simple)) (PP (IN in) (NP (DT the) (NNP Wood) (NNP Memorial))) (PP (IN on) (NP (NNP Saturday))) (PP (IN at) (NP (NNP Aqueduct))))))) (, ,) (CC but) (VP (VBD said) (NP (NNP Wednesday)) (SBAR (S (NP (PRP he)) (VP (VBZ considers) (S (NP (DT A) (NNP P) (NNP Valentine)) (VP (TO to) (VP (VB be) (NP (PRP$ his) (JJ top) (NN 3-year-old)))))))))) (. .)))
</P>
<P>
A P Valentine 's rivals in the $ 750,000 Blue Grass are expected to be Hero 's Tribute , this year an allowance winner and a troubled third in the Louisiana Derby ; Dollar Bill , winner of the Risen Star and fourth in the Louisiana Derby after almost falling ; Invisible Ink , third in the Florida Derby ; Millennium Wind , the Santa Catalina winner and second in the Louisiana Derby ; Bonnie Scott , winner of four of six starts at Turfway Park ; and Songandaprayer , winner of the Fountain of Youth , then fifth in the Florida Derby .
</P>
<P>
( (S (NP (NNP Songandaprayer)) (VP (AUXZ is) (VP (VBN co-owned) (PP (IN by) (NP (NP (NNP Bobby) (NNP Hurley)) (, ,) (NP (NP (NP (DT an) (JJ All-American) (NN guard)) (PP (IN at) (NP (NNP Duke)))) (CC and) (NP (DT a) (JJ former) (NNP NBA) (NN player))))))) (. .)))
</P>
```